### PR TITLE
Chapter4

### PR DESCRIPTION
**Pull request by DerekAmuna**
# Chapter 4 deployment methods
Please do let me know when chapter 3 pr and this pr are approved so i can rebase before merging to avoid conflicts again

### CODE REVIEW

**Review by emmanuelkyeremeh — APPROVED** [review]

**Review by emmanuelkyeremeh — APPROVED** [review]

File: chapter4/webservice/intro_lm.bin
Comment (Silas-Asamoah) [inline chapter4/webservice/intro_lm.bin:1]: This file should be gitignored

File: chapter4/webservice/dv.bin
Comment (Silas-Asamoah) [inline chapter4/webservice/dv.bin:1]: This file should be gitignored

File: chapter4/webservice/mlflow_artifacts/1/64b1755f573c4439974d9f5ab58a9a06/artifacts/dv.bin
Comment (Silas-Asamoah) [inline chapter4/webservice/mlflow_artifacts/1/64b1755f573c4439974d9f5ab58a9a06/artifacts/dv.bin:1]: This file should be gitignored

File: chapter4/webservice/mlflow_artifacts/dv.bin
Comment (Silas-Asamoah) [inline chapter4/webservice/mlflow_artifacts/dv.bin:1]: This file should be gitignored

File: chapter4/webservice/predict.py [hunk 1]
```diff
@@ -0,0 +1,63 @@
+# service without mlflow server
+from flask import Flask, request, jsonify
+import pickle
+import mlflow
+from mlflow.tracking import MlflowClient
+
+
+MLFLOW_TRACKING_URI = "http://127.0.0.1:5000"
```
Comment (Silas-Asamoah) [inline chapter4/webservice/predict.py]: This should be read from an environment variable and not hard coded

File: chapter4/webservice/predict2.py [hunk 2]
```diff
@@ -0,0 +1,58 @@
+# service with mlflow server and model saved in mlflow server with pipeline
+from flask import Flask, request, jsonify
+import mlflow
+import numpy as np
+
+
+MLFLOW_TRACKING_URI = "http://127.0.0.1:5000"
```
Comment (Silas-Asamoah) [inline chapter4/webservice/predict2.py]: Likewise here

File: chapter4/webservice/predict3.py [hunk 3]
```diff
@@ -0,0 +1,58 @@
+# batch deployment
+from flask import Flask, request, jsonify
+import mlflow
+import numpy as np
+
+
+MLFLOW_TRACKING_URI = "http://127.0.0.1:5000"
```
Comment (Silas-Asamoah) [inline chapter4/webservice/predict3.py]: Here as well

File: homework/chapter4/intro_lm.bin
Comment (Silas-Asamoah) [inline homework/chapter4/intro_lm.bin:1]: This file should be gitignored

**Review by Silas-Asamoah — CHANGES_REQUESTED** [review]
Left some inline comments , kindly check it out

File: homework/chapter4/dockerfile [hunk 4]
```diff
@@ -1,5 +1,7 @@
 FROM agrigorev/zoomcamp-model:mlops-2024-3.10.13-slim
 
+ENV MLFLOW_TRACKING_URI=http://127.0.0.1:5000
```
Comment (Silas-Asamoah) [inline homework/chapter4/dockerfile]: This is still hardcoded and shouldn't be in your Dockerfile. Create a `.env`(this should be gitignored) and read the tracking URI from there. Here's a StackOverflow link on this 
https://stackoverflow.com/questions/46917831/how-to-load-several-environment-variables-without-cluttering-the-dockerfile/46925953

**Review by Silas-Asamoah — CHANGES_REQUESTED** [review]
Great job on the comments, just one last thing on the environment variables with Docker and this PR should be good to go

**Review by Silas-Asamoah — APPROVED** [review]

**Pull request merged**